### PR TITLE
Add a dataset class for loading Tensorflow Datasets

### PR DESCRIPTION
**Pull request by tfujiwar**
## Motivation and Context
To be able to run training with datasets prepared in TensorFlow Datasets.

## Description
Implemented a dataset class for loading TensorFlow Datasets for classification.

## How has this been tested?
Executed several trainings with different `train_path` options.
- `gs://lmfs-backend-us-west1/home/sawada/tensorflow_datasets/mnist`
- `/home/blueoil/tensorflow_datasets/mnist`
- `/home/blueoil/tensorflow_datasets/cifar10`
- `/home/blueoil/tensorflow_datasets/cifar100`
- `/home/blueoil/tensorflow_datasets/fashion_mnist`
- `/home/blueoil/tensorflow_datasets/tf_flowers`
- `/home/blueoil/tensorflow_datasets/rock_paper_scissors`

Sample config file:
```yaml
task_type: classification

network_name: LmnetV1Quantize

dataset:
  format: TensorFlow Datasets for Classification
  train_path: gs://lmfs-backend-us-west1/home/sawada/tensorflow_datasets/mnist
  test_path:

trainer:
  batch_size: 10
  epochs: 20
  optimizer: Adam
  learning_rate_schedule: constant
  initial_learning_rate: 0.01

network:
  quantize_first_convolution: yes

common:
  image_size:
    - 32  # height
    - 32  # width

  pretrain_model: false
  dataset_prefetch: true
```

Please be sure to be able to access the GCS bucket. For example, you need to generate a credential file by `gcloud` command before executing training.

```sh
gcloud auth application-default login
python blueoil/blueoil_train.py -c /home/blueoil/config/test.yml -i test
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
@tfujiwar Thanks!! I'll check soon.

File: lmnet/lmnet/datasets/tensorflow_datasets.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    """
+    @property
+    def classes(self):
+        return self.info.features["label"].names
+
+    @property
+    def num_classes(self):
+        return self.info.features["label"].num_classes
+
+    def _init_features(self):
```
Comment (ruimashita) [inline lmnet/lmnet/datasets/tensorflow_datasets.py]: Do `self.images` and `self.labels` have all of data on memory?
It is better to get a data for each `__getitem__` for big dataset.

File: lmnet/lmnet/datasets/tensorflow_datasets.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+from lmnet import data_processor
+
+
+class TensorFlowDatasetsBase(Base):
+    """
+    Abstract dataset class for loading TensorFlow Datasets.
+    The parameter "extend_dir" should be a path to the prepared data of tfds.
+    "extend_dir" can be a path either in local or in Google Cloud Storage.
+    """
+    available_subsets = ["train", "train_validation_saving", "validation"]
```
Comment (ruimashita) [inline lmnet/lmnet/datasets/tensorflow_datasets.py]: Why do you fix `available_subsets` as `["train", "train_validation_saving", "validation"]` ? This doesn't include `test` subset.

File: lmnet/lmnet/datasets/tensorflow_datasets.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Initializing feature variables by using tf.Tensor from datasets.
+        For classification, "images" and "labels" should be initialized.
+        For object detection, "images" and "gt_boxes" should be initialized.
+        """
+        raise NotImplementedError()
+
+    @property
+    def num_per_epoch(self):
+        return len(self.images)
```
Comment (ruimashita) [inline lmnet/lmnet/datasets/tensorflow_datasets.py]: Maybe `builder.info` has the information of data number.

File: lmnet/lmnet/datasets/tensorflow_datasets.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+
+class TensorFlowDatasetsClassification(TensorFlowDatasetsBase):
```
Comment (ruimashita) [inline lmnet/lmnet/datasets/tensorflow_datasets.py]: Please write what the class require tfds's `features` structure (maybe `label` and `image` ), addition this, It is better to check the reuirements in code.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 1 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: Yes, they have all the data on memory. I agree that it's not good for big datasets but there is no way to access TensorFlow Datasets by index because TensorFlow Datasets work as iterators. So current implementation is a workaround. To solve this problem, we need to implement this class as a new dataset iterator.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 2 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: Thanks. I'll fix it.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 4 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: Thank you. I'll add a comment and codes to check it.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 1 above)
Comment (tkng) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: Another idea is tha we can ignore index in the ```__getitem__``` method.

e.g. load 10000 data into memory (from TF datasets iterator), shuffle, then always pop the 1st item in the list in ```__getitem__``` method.

That's not a straightforward implementation, but as a tentative solution, we can tolerate it.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 3 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: The number of data in pre-defined split (e.g. TRAIN, VALIDATION, TEST) is included in `builder.info`. But TRAIN split will be splitted into two sub-splits if datasets only have TRAIN split.
https://github.com/tfujiwar/blueoil/blob/828d152d7f5d04cda8be9000ff9b46e7847b3d4b/lmnet/lmnet/datasets/tensorflow_datasets.py#L77-L80

That's why `len(self.images)` is used here.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 2 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: This variable is defined as `@staticmethod` in the base class. But available subsets will be decided after the class is instantiated in case of this class because it depends on what dataset will be loaded.

So I'll define this variable as `["train", "validation"]` in this class because these two subsets are always available.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 1 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: @tkng
Thanks for your suggestion. I think it's a good idea.
Currently, the method `__getitem__` seems to be used only for shuffling in dataset_iterator.

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 4 above)
Comment (ruimashita) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: 👍

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 2 above)
Comment (ruimashita) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: > So I'll define this variable as ["train", "validation"] in this class because these two subsets are always available.

Nice 👍

File: lmnet/lmnet/datasets/tensorflow_datasets.py (hunk 1 above)
Comment (tfujiwar) [reply lmnet/lmnet/datasets/tensorflow_datasets.py]: I tried to update `__getitem__` as we talked above. I used `tf.Session().run()` for getting next element but I found that it doesn't work with the option `enable_prefetch = True`. Maybe we can fix this problem by changing `DatasetIterator` in order to use tf.Session().run() in multi processes safely.

https://github.com/blue-oil/blueoil/blob/cf5c085933aba3b7d7f319468518fd557b1322ff/lmnet/lmnet/datasets/tensorflow_datasets.py#L62-L67

But I think we should not change many things only for introducing TensorFlowDatasets class because this class is a temporary solution anyway. I and @ruimashita talked and decided not to solve the problem this time. All the data will be loaded on initialization of the dataset class.
